src/circuit/gadget/sinsemilla/chip/hash_to_point.rs

```diff
@@ -0,0 +1,447 @@
+            // multiplied by zero.)
+            {
+                region.assign_advice(
+                    || "dummy lambda2",
+                    config.lambda_2,
+                    offset,
+                    || Ok(pallas::Base::zero()),
+                )?;
+                region.assign_advice(
+                    || "dummy x_p",
+                    config.x_p,
+                    offset,
+                    || Ok(pallas::Base::zero()),
+                )?;
+            }
+
+            CellValue::new(y_a_cell, y_a.0)
+        };
+
+        #[cfg(test)]
+        #[allow(non_snake_case)]
+        // Check equivalence to result from primitives::sinsemilla::hash_to_point
+        {
+            use crate::circuit::gadget::sinsemilla::message::MessagePiece;
+            use crate::primitives::sinsemilla::{K, S_PERSONALIZATION};
+            use group::prime::PrimeCurveAffine;
+            use pasta_curves::arithmetic::CurveExt;
+
+            let field_elems: Option<Vec<pallas::Base>> =
+                message.iter().map(|piece| piece.field_elem()).collect();
+
+            if field_elems.is_some() {
+                // Get message as a bitstring.
+                let bitstring: Vec<bool> = message
+                    .iter()
+                    .map(|piece: &MessagePiece<pallas::Base, K>| {
+                        piece
+                            .field_elem()
+                            .unwrap()
+                            .to_le_bits()
+                            .into_iter()
+                            .take(K * piece.num_words())
+                            .collect::<Vec<_>>()
+                    })
+                    .flatten()
+                    .collect();
+
+                let hasher_S = pallas::Point::hash_to_curve(S_PERSONALIZATION);
+                let S = |chunk: &[bool]| hasher_S(&lebs2ip_k(chunk).to_le_bytes());
+
+                // We can use complete addition here because it differs from
+                // incomplete addition with negligible probability.
+                let expected_point = bitstring
+                    .chunks(K)
+                    .fold(Q.to_curve(), |acc, chunk| (acc + S(chunk)) + acc);
+                let actual_point =
+                    pallas::Affine::from_xy(x_a.value().unwrap(), y_a.value().unwrap()).unwrap();
+                assert_eq!(expected_point.to_affine(), actual_point);
+            }
+        }
+
+        Ok((EccPoint::from_coordinates_unchecked(x_a.0, y_a), zs_sum))
+    }
+
+    #[allow(clippy::type_complexity)]
+    // Hash a message piece containing `piece.length` number of `K`-bit words.
+    // To avoid a duplicate assignment, the accumulator x-coordinate provided
+    // by the caller is not copied. This only works because `hash_piece()` is
+    // an internal API. Before this call to `hash_piece()`, x_a MUST have been
+    // already assigned within this region at the correct offset.
```

Comment:
```suggestion
    /// Hashes a message piece containing `piece.length` number of `K`-bit words.
    ///
    /// To avoid a duplicate assignment, the accumulator x-coordinate provided
    /// by the caller is not copied. This only works because `hash_piece()` is
    /// an internal API. Before this call to `hash_piece()`, x_a MUST have been
    /// already assigned within this region at the correct offset.
```